message/batch.go

```diff
@@ -89,6 +113,28 @@ type BatchChild struct {
 	Values []*primitive.Value
 }
 
+func (c *BatchChild) Clone() *BatchChild {
+	var newQueryOrId interface{}
+	if c.QueryOrId != nil {
+		switch queryOrId := c.QueryOrId.(type) {
+		case string:
+			newQueryOrId = queryOrId
+		case []byte:
+			newQueryOrId = primitive.CloneByteSlice(queryOrId)
+		default:
+			log.Error().Msgf("could not clone QueryOrId, unexpected type: %T", queryOrId)
```

Comment:
Changing the `Clone()` signature because of this edge case seems a bit overkill. In theory this should never happen but we are relying on the callers to enforce this restriction at the moment. The ideal implementation would prevent callers from modifying this field directly so that an error could be thrown if an attempt to create a  `BatchChild` with an invalid `QueryOrId` was made.

Reply:
This `Clone()` method is specific to `BatchChild`, it doesn't seem overkill to me that it could return 2 arguments, you could even call it `MaybeClone()` if you prefer.

> In theory this should never happen but we are relying on the callers to enforce this restriction at the moment. The ideal implementation would prevent callers from modifying this field directly.

If it should never happen, then don't protect against it. I didn't create this project with such a defensive approach in mind. Echoing our offline discussion, the idea was more of a "bare-bones" data structure + the assumption that users know how to use it properly.

Reply:
>>> If it should never happen, then don't protect against it. I didn't create this project with such a defensive approach in mind. Echoing our offline discussion, the idea was more of a "bare-bones" data structure + the assumption that users know how to use it properly.

Makes sense